kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Configuration.java

```diff
@@ -117,43 +92,29 @@ private static void checkNamedFiltersAreDefined(Set<String> filterDefsByName,
             if (!duplicatedNames.isEmpty()) {
                 throw new IllegalConfigurationException("'filterDefinitions' contains multiple items with the same names: " + duplicatedNames);
             }
-        }
-
-        // Enforce post condition: Every filter referenced by a name is defined in the filterDefinitions
-        Set<String> filterDefsByName = Optional.ofNullable(filterDefinitions).orElse(List.of()).stream().map(NamedFilterDefinition::name).collect(
-                Collectors.toSet());
-        checkNamedFiltersAreDefined(filterDefsByName, defaultFilters, "defaultFilters");
-        for (var virtualCluster : virtualClusters) {
-            checkNamedFiltersAreDefined(filterDefsByName, virtualCluster.filters(), "virtualClusters." + virtualCluster.name() + ".filters");
-        }
 
-        // Every filter defined in the filterDefinitions is used somewhere
-        if (filterDefinitions != null) {
-            var defined = filterDefinitions.stream().map(NamedFilterDefinition::name).collect(Collectors.toCollection(HashSet::new));
-            if (defaultFilters != null) {
-                defaultFilters.forEach(defined::remove);
+            // Enforce post condition: Every filter referenced by a name is defined in the filterDefinitions
+            Set<String> filterDefsByName = Optional.ofNullable(filterDefinitions).orElse(List.of()).stream().map(NamedFilterDefinition::name).collect(
+                    Collectors.toSet());
+            checkNamedFiltersAreDefined(filterDefsByName, defaultFilters, "defaultFilters");
+            for (var virtualCluster : virtualClusters) {
+                checkNamedFiltersAreDefined(filterDefsByName, virtualCluster.filters(), "virtualClusters." + virtualCluster.name() + ".filters");
             }
-            virtualClusters.stream()
-                    .map(VirtualCluster::filters)
-                    .filter(Objects::nonNull)
-                    .flatMap(Collection::stream)
-                    .forEach(defined::remove);
-            if (!defined.isEmpty()) {
-                throw new IllegalConfigurationException(
-                        "'filterDefinitions' defines filters which are not used in 'defaultFilters' or in any virtual cluster's 'filters': " + defined);
-            }
-        }
 
-        if (filters != null && virtualClusters.stream()
-                .map(VirtualCluster::filters)
-                .anyMatch(Objects::nonNull)) {
-            throw new IllegalConfigurationException(
-                    "'filters' cannot be specified on a virtual cluster when 'filters' is defined at the top level.");
+            checkAllNamedFilterAreUsed(filterDefinitions, virtualClusters, defaultFilters);
```

Comment:
I'm on the fence about this one. It feels a bit over zealous really. In a (operator) managed environment it should never trigger in an unmanged environment I wonder if it causes more trouble than it saves? Developer copies prod config locally and comments out the reference to audit logging filter so his local testing doesn't get logged. Do we really want the proxy to refuse to start?

I understand the case we are trying to defend against is typo in production failing to wire together. Maybe we need to think about some sort of environment awareness...